SU2_CFD/src/solvers/CEulerSolver.cpp

```diff
@@ -845,7 +845,8 @@ void CEulerSolver::Set_MPI_Nearfield(CGeometry *geometry, CConfig *config) {
       /*--- Communicate the counts to iDomain with non-blocking sends ---*/
 
       SU2_MPI::Isend(&nPointTotal_s[iDomain], 1, MPI_UNSIGNED_LONG, iDomain, iDomain, SU2_MPI::GetComm(), &req);
-      SU2_MPI::Request_free(&req);
+      SU2_MPI::Wait(&req,&status);
+      //SU2_MPI::Request_free(&req);
```

Comment:
Have you stress tested that this does not create deadlocks?

Reply:
No, but `SU2_MPI::Request_free` creates errors shown below during adjoint calculation, so `SU2_MPI::Wait` was a workaround. Is there a better way?

Error in function AMPI_Request_free (../externals/medi/include/medi/ampi/async.hpp:249)
The message is: Freeing a handle that is not finish with wait, waitall, etc..

Reply:
I see... That routine was written to use non-blocking comms (whether it needs that or not is a different story).
By waiting right after the Send/Recv you make it equivalent to a blocking communication.
If the non-blocking was indeed necessary this may cause the code to dead-lock because a Send will not be matched with the right Recv.
It would be good to test this with gross-oversubscribtion, e.g. 17 ranks on a system with 4 cores while playing a youtube video.

Reply:
I already encountered an issue without doing the test you mentioned. I changed a number of cores to use and the adjoint calculation seems to stop there. As I'm not very familiar with MPI, I was doing some research on MPI but I have not been able to solve this so far.

Reply:
Ok. Based on CGeometry::PostP2PRecvs (and similar).
You need to create a vector of all the requests (one per message).
You launch all the IRecv first.
Then you launch all the sends.
Finally you "Waitall" on the receive requests.

Reply:
I tried to write something similar to `CGeometry::PostP2PRecvs`, `CGeometry::PostP2PSends`, and `CGeometry::CompleteComms` but I cannot get it to work for some reason. Segmentation fault occurs during Waitall.  Do you have any idea on what is possible causing this?

```
*** Process received signal ***
Signal: Segmentation fault (11)
Signal code: Address not mapped (1)
Failing at address: 0x1520
*** An error occurred in MPI_Waitall
*** reported by process [2490630145,3]
*** on communicator MPI_COMM_WORLD
*** MPI_ERR_REQUEST: invalid request
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
```

Reply:
I will try to help when I have some time.

Reply:
Thanks. I'm still working on it (slowly).

Reply:
After learning how MPI works, MPI appears to be working now but the program stops with the following messages. Google says it is related to memory issues where you delete the same pointer twice or you try to access address outside of a vector, but I cannot find such issues in my codes. Can you help me identify an issue?

```
*** Process received signal ***
Signal: Aborted (6)
Associated errno: Unknown error 32766 (32766)
Signal code:  (-543634320)
```
and
```
free(): invalid next size (fast)
```

Reply:
You can try using valgrind. --buildtype=debugoptimized, then mpirun -n "n" valgrind SU2_CFD ... and you should get messages if you are double-freeing or accessing out of bounds. But if it is happening inside MPI it will be difficult.

That entire function was really badly written, it should be doing an AllToAll for the counts followed by an AllToAllV for the data, not all this complication with Sends and Recvs.

Reply:
Thanks! valgrind found an issue and I fixed it. It seems to be working now. I have introduced some variables for debug purpose, so I will do a little clean-up later.

Reply:
I think I'm now done with this comments.